dm/syncer/compactor.go

```diff
@@ -192,34 +190,30 @@ func (c *compactor) compactJob(j *job) {
 	prevJob := c.buffer[prevPos]
 	c.logger.Debug("start to compact", zap.Stringer("previous dml", prevJob.dml), zap.Stringer("current dml", j.dml))
 
-	switch j.tp {
-	case update:
-		if prevJob.tp == insert {
-			// INSERT + UPDATE => INSERT
-			j.tp = insert
-			j.dml.oldValues = nil
-			j.dml.originOldValues = nil
-			j.dml.op = insert
+	// adjust safemode
+	skipReduce := false
+	switch j.dml.Type() {
+	case sqlmodel.RowChangeUpdate:
+		if prevJob.dml.Type() == sqlmodel.RowChangeInsert {
 			// DELETE + INSERT + UPDATE => INSERT with safemode
-			j.dml.safeMode = prevJob.dml.safeMode
-		} else if prevJob.tp == update {
-			// UPDATE + UPDATE => UPDATE
-			j.dml.oldValues = prevJob.dml.oldValues
-			j.dml.originOldValues = prevJob.dml.originOldValues
+			j.safeMode = prevJob.safeMode
 		}
-	case insert:
-		if prevJob.tp == del {
+	case sqlmodel.RowChangeInsert:
+		if prevJob.dml.Type() == sqlmodel.RowChangeDelete {
 			// DELETE + INSERT => INSERT with safemode
-			j.dml.safeMode = true
+			j.safeMode = true
+			skipReduce = true
 		}
-	case del:
-		// do nothing because anything + DELETE => DELETE
+	}
```

Comment:
Why not also skip reduce RowChangeDelete?

Reply:
we only run compactor for row changes whose PR/UK doesn't change, I think it's both OK. 

And I want to apply Reduce to all combinations of row change so I want to reduce the presence of `skipReduce`. If we implement multi-row version of UPDATE in future, I think we can also remove the `// DELETE + INSERT => INSERT with safemode` extra path